external/avro/src/main/scala/org/apache/spark/sql/avro/AvroEncoder.scala

```diff
@@ -0,0 +1,534 @@
+              If(IsNull(fieldValue), tree, fieldValue)
+          }
+        } else {
+          deserializerFor(resolvedSchema, path)
+        }
+
+      case RECORD =>
+        val args = avroSchema.getFields.asScala.map { field =>
+          val position = Literal(field.pos)
+          val argument = deserializerFor(field.schema, addToPath(field.name))
+          (position, argument)
+        }.toList
+        val recordClass = Option(ReflectData.get.getClass(avroSchema))
+          .getOrElse(classOf[GenericData.Record])
+        val newInstance = if (recordClass == classOf[GenericData.Record]) {
+          NewInstance(
+            recordClass,
+            Invoke(
+              Literal.fromObject(
+                new SerializableSchema(avroSchema),
+                ObjectType(classOf[SerializableSchema])),
+              "value",
+              ObjectType(classOf[Schema]),
+              Nil) :: Nil,
+            ObjectType(recordClass))
+        } else {
+          NewInstance(
+            recordClass,
+            Nil,
+            ObjectType(recordClass))
+        }
+        val result = InitializeAvroObject(newInstance, args)
+
+        If(IsNull(path), Literal.create(null, ObjectType(recordClass)), result)
+
+      case NULL =>
+        // Encountering NULL at this level implies it was the type of a Field, which should never
+        // be the case.
+        throw new IncompatibleSchemaException("Null type should only be used in Union types")
+    }
+  }
+
+  /**
+   * Returns an expression that can be used to serialize an Avro object with a class of type `T`
+   * that is compatible with the given Schema to an InternalRow
+   */
+  def serializerFor[T <: IndexedRecord](
+      avroClass: Class[T], avroSchema: Schema): Expression = {
+    val inputObject = BoundReference(0, ObjectType(avroClass), nullable = true)
+    val nullSafeInput = AssertNotNull(inputObject, Seq("top level"))
+    serializerFor(nullSafeInput, avroSchema)
+  }
+
+  def serializerFor(
+      inputObject: Expression,
+      avroSchema: Schema): Expression = {
+
+    def toCatalystArray(inputObject: Expression, schema: Schema): Expression = {
+      val elementType = inferExternalType(schema)
+      if (elementType.isInstanceOf[ObjectType]) {
+        MapObjects(element =>
+          serializerFor(element, schema),
+          Invoke(
+            inputObject,
+            "toArray",
+            ObjectType(classOf[Array[Object]])),
+          elementType)
+      } else {
+        NewInstance(
+          classOf[GenericArrayData],
+          inputObject :: Nil,
+          dataType = ArrayType(elementType, containsNull = false))
+      }
+    }
+
+    def toCatalystMap(inputObject: Expression, schema: Schema): Expression = {
+      val valueSchema = schema.getValueType
+      val valueType = inferExternalType(valueSchema)
+      ExternalMapToCatalyst(
+        inputObject,
+        ObjectType(classOf[org.apache.avro.util.Utf8]),
```

Comment:
In the previous [databricks/spark-avro#217](https://github.com/databricks/spark-avro/pull/217), user @mtraynham points out we should use `CharSequence` as the object type:

```suggestion
        ObjectType(classOf[java.lang.CharSequence]),
```

See [comment](https://github.com/databricks/spark-avro/pull/217#issuecomment-405305815) and [suggestion](https://github.com/databricks/spark-avro/pull/217#issuecomment-405385022).

Reply:
Ah, thanks for reminding, I missed this.